third_party/move/move-compiler-v2/src/bytecode_generator.rs

```diff
@@ -822,14 +820,29 @@ impl<'env> Generator<'env> {
         }
     }
 
-    fn gen_arg_list(&mut self, exps: &[Exp]) -> Vec<TempIndex> {
-        exps.iter().map(|exp| self.gen_arg(exp)).collect()
+    /// Generate the code for a list of arguments.
+    fn gen_arg_list(&mut self, exps: &[Exp], force_temp: bool) -> Vec<TempIndex> {
+        let len = exps.len();
+        // Generate code with forced creation of temporaries for all except last arg.
+        let mut args = exps
+            .iter()
+            .take(if len == 0 { 0 } else { len - 1 })
+            .map(|exp| self.gen_arg(exp, force_temp))
+            .collect::<Vec<_>>();
+        // If there is a last arg, we don't need to force create a temporary for it.
+        if let Some(last_arg) = exps.iter().last().map(|exp| self.gen_arg(exp, false)) {
+            args.push(last_arg);
+        }
+        args
     }
 
-    fn gen_arg(&mut self, exp: &Exp) -> TempIndex {
+    /// Generate the code for an argument.
+    /// If `with_forced_temp` is true, it will force generating a temporary for the argument,
+    /// thereby forcing its evaluation right away in the generated code.
+    fn gen_arg(&mut self, exp: &Exp, with_forced_temp: bool) -> TempIndex {
         match exp.as_ref() {
-            ExpData::Temporary(_, temp) => *temp,
-            ExpData::LocalVar(id, sym) => self.find_local(*id, *sym),
+            ExpData::Temporary(_, temp) if !with_forced_temp => *temp,
```

Comment:
Actually I'm not quite understanding this. If the exp is already a temp or a local, we shouldn't need to introduce another temp/local for evaluaton order, since the expression is already evaluated?

Reply:
yeah, for instance, in this statement, if i is passed as a function argument, a temp is generated for it `vector::swap(v, i, { i = i + 1; i })` , if we don't evaluate it first, i = i + 1 will be executed

Reply:
Ah thanks, now remembering the issue. Lets hope that copy propagation will eliminate all those many temps...